File: sum_path.cpp
//TC: O(n) where n is number of tree elements
//SC: O(n) since the recursive stack grows till n elements



/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int result=0;
    
    void preorder(TreeNode* root, int currval){
        
        if(root == NULL)
            return;
        
        currval = currval*10 + root->val;
        
        if(root->left == NULL && root->right == NULL)      
            result += currval;
        
        preorder(root->left, currval);
        preorder(root->right, currval);
        
        
    }
    
    
    int sumNumbers(TreeNode* root) {
        
        //can use a helper function to perform pre-order traversal using recursion
        int currval = 0;
        preorder(root, currval);
        return result;        
        
    }
    
    
};

File: sumpath.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def treePathsSumUtil(self,root, val): 
  
        # Base Case 
        if root is None: 
            return 0

        # Update val 
        val = (val*10 + root.val) 

        # If current node is leaf, return the current value of val 
        if root.left is None and root.right is None: 
            return val 

        # Recur sum of values for left and right subtree 
        return (self.treePathsSumUtil(root.left, val) + 
                self.treePathsSumUtil(root.right, val)) 
    
    def sumNumbers(self, root: TreeNode) -> int:
        if root is None:
            return 0
        if root.left is None and root.right is None:
            return root.val
        return self.treePathsSumUtil(root.left, root.val) + self.treePathsSumUtil(root.right, root.val)
    
